File: 3.dart
import 'dart:collection';
import 'dart:math';

class Solution {
  int lengthOfLongestSubstring(String s) {
    final indexMap = HashMap<String, int>();
    var res = 0, j = 0;

    for (var i = 0; i < s.length; i++) {
      if (indexMap.containsKey(s[i])) {
        res = max(res, i - j);
        j = max(indexMap[s[i]]! + 1, j);
      }
      indexMap[s[i]] = i;
    }
    return max(res, s.length - j);
  }
}

File: 3.cpp
#include <algorithm>
#include <ranges>
#include <string>
#include <unordered_set>

class Solution {
public:
  int lengthOfLongestSubstring(const std::string s) {
    std::unordered_set<char> substring;
    auto maxSize = 0;
    auto i = 0, j = 0;

    while (i < std::ranges::ssize(s)) {
      auto current = s[i];

      if (substring.contains(current)) {
        maxSize = std::max(maxSize, i - j);
		
        while (s[j] != current)
          substring.erase(s[j++]);
        substring.erase(s[j++]);
      }

      substring.insert(current);
      i++;
    }

    return std::max(maxSize, i - j);
  }
};